web-common/src/features/entity-management/file-artifacts-store.ts

```diff
@@ -1,131 +1,241 @@
+    const curName = get(this.name);
+    if (
+      curName?.name !== resource.meta?.name?.name ||
+      curName?.kind !== resource.meta?.name?.kind
+    ) {
+      this.name.set({
+        kind: resource.meta?.name?.kind,
+        name: resource.meta?.name?.name,
+      });
+    }
+    this.lastStateUpdatedOn = resource.meta?.stateUpdatedOn;
+  }
+
+  public deleteResource() {
+    this.name.set(undefined);
+  }
+}
+
+export class FileArtifactsStore {
+  /**
+   * Map of all files and whether it is reconciling or not.
+   * If an entry is present here then there should be on in {@link artifacts} as well
+   */
+  public readonly files = writable<Record<string, boolean>>({});
+
+  /**
+   * Map of all files and its individual store
+   */
+  private readonly artifacts: Record<string, FileArtifact> = {};
+
+  // Actions
+
+  public async init(instanceId: string) {
+    const resourcesResp = await runtimeServiceListResources(instanceId);
+    if (!resourcesResp.resources?.length) return;
+    for (const resource of resourcesResp.resources) {
+      switch (resource.meta?.name?.kind) {
+        case ResourceKind.Source:
+        case ResourceKind.Model:
+        case ResourceKind.MetricsView:
+        case ResourceKind.Chart:
+        case ResourceKind.Dashboard:
+          this.setResource(resource);
+          break;
+      }
     }
-    callback(state.entities[path]);
-    return state;
-  });
-};
-
-const fileArtifactsEntitiesReducers = {
-  setName(path: string, name: string) {
-    createOrUpdateFileArtifact(path, (entityData) => {
-      entityData.name = name;
+  }
+
+  public updateFile(filePath: string) {
+    if (get(this.files)[filePath]) return;
+    this.artifacts[filePath] ??= new FileArtifact(filePath);
+    this.files.update((f) => {
+      f[filePath] = false;
+      return f;
     });
-  },
+  }
+
+  public deleteFile(filePath: string) {
+    delete this.artifacts[filePath];
+    this.files.update((f) => {
+      delete f[filePath];
+      return f;
+    });
+  }
+
+  public setResource(resource: V1Resource) {
+    this.updateArtifact(resource);
+    this.updateReconciling(resource);
+  }
 
-  setErrors(affectedPaths: Array<string>, errors: Array<V1ReconcileError>) {
-    const errorsForPaths = new Map<string, Array<V1ReconcileError>>();
-    affectedPaths.forEach((affectedPath) =>
-      errorsForPaths.set(correctFilePath(affectedPath), []),
+  public updateReconciling(resource: V1Resource) {
+    this.files.update((f) => {
+      resource.meta?.filePaths?.map(correctFilePath).forEach((filePath) => {
+        f[filePath] =
+          resource.meta?.reconcileStatus ===
+          V1ReconcileStatus.RECONCILE_STATUS_RUNNING;
+      });
+      return f;
+    });
+  }
+
+  public updateArtifact(resource: V1Resource) {
+    resource.meta?.filePaths?.map(correctFilePath).forEach((filePath) => {
+      this.artifacts[filePath] ??= new FileArtifact(filePath);
+      this.artifacts[filePath].updateResource(resource);
+    });
+  }
```

Comment:
I was wondering why these methods are on the `FileArtifactsStore`, rather than on the `FileArtifact` class itself. Then I realized that each update updates not just one artifact, but potentially >1.

So improved naming could be the plural `updateArtifacts` (and will assume `isReconciling` will be updated as a part of that function). Further, `updateFileArtifactsForResource` could be most clear.

Reply:
Yeah I get that. It just took me a little while to understand what was going on. The plural `updateArtifacts` is good.